### PR TITLE
Added Edit Notebook Metadata Option (#6402)

### LINKED ISSUE #6402: Edit notebook metadata
### Problem

In the classic notebook it's possible to edit the notebook metadata:

image

image

### Proposed Solution

We add a new plugin to replicate that UX and let users edit notebook metadata in a dialog.

### Additional context

In JupyterLab the notebook tools is in the right area:

image

### PR DESCRIPTION
**Pull request by Dilip-Jain**
### Problem
Feature Required: Edit Notebook Metadata option under the Edit Menu (in Notebook) as was in classic notebook.
@jtpio mentions -
> Maybe the simplest way for now would be to add a Edit Notebook Metadata command to the menu that would open the Notebook tools panel instead of a dialog.

### Solution
Edit Notebook Metadata plugin added under Edit Menu for functionality similar to classic-notebook. Clicking on the option under the menu opens up the Notebook Tools widget in the right sidebar and expands the Additional Tools (which contains the Notebook Metadata editor) collapsible section (default: collapsed).
**Fixes #6402** 

### Key Addition [Enhancement]
`editNotebookMetadata` plugin to add a command to open right sidebar for Editing Notebook Metadata when clicking on "Edit Notebook Metadata" option under Edit menu.
- On the click, the plugin first checks if Show Notebook Tools (Right Sidebar) is open (expanded)
- If not open, toggle the Right Sidebar and expand the Show Notebook Tools.
- Lastly, it expands the 'Advanced Tools' collapsible section (default, collapsed)

Thank you for considering this pull request, and I welcome any feedback or suggestions for further improvements.

### CODE REVIEW

File: packages/notebook-extension/src/index.ts [hunk 1]
```diff
@@ -478,12 +478,65 @@ const trusted: JupyterFrontEndPlugin<void> = {
   },
 };
 
+/**
+ * Add a command to open right sidebar for Editing Notebook Metadata when clicking on "Edit Notebook Metadata" under Edit menu
+ */
+const editNotebookMetadata: JupyterFrontEndPlugin<void> = {
+  id: '@jupyter-notebook/notebook-extension:edit-notebook-metadata',
```
Comment (jtpio) [inline packages/notebook-extension/src/index.ts:485]: @Dilip-Jain curious if you tried adding the command to the menu settings, as mentioned in https://github.com/jupyter/notebook/issues/6402#issuecomment-1743087342 ?

To know if the plugin is necessary, or if the menu entry would be enough.

Comment (Dilip-Jain) [reply packages/notebook-extension/src/index.ts:485]: For enhancing the "Edit Notebook Metadata" feature, adhering to the user experience, the objective was to open both the right sidebar (showing the "Notebook Tools" widget) and automatically expand the "Additional Tools" collapsible section, which houses the notebook metadata editor. This streamlined action ensures a smoother and more intuitive user interaction.

The desired action involves two key steps:

1. **Opening the Right Sidebar:** This step involves displaying the "Notebook Tools" widget, which is housed in the right sidebar.
2. **Expanding the "Additional Tools" Section:** Within the right sidebar, the "Additional Tools" collapsible section contains the notebook metadata editor. This section needs to be automatically expanded to provide immediate access to the editor.

The challenge faced was that using the existing command `application:toggle-panel` with specified arguments in the menus.json settings file accomplished the first step—opening the right sidebar. However, it did not automatically expand the "Additional Tools" section to reveal the notebook metadata editor so to overcome this limitation a custom plugin was added.

Screenshot: Right Sidebar panel:
**Using command menu entry:**
image

**Using the plugin:**
image

**Comment by Dilip-Jain** [comment]
Hey @jtpio, 
anything not in place or needs to be altered?

File: packages/notebook-extension/src/index.ts [hunk 2]
```diff
@@ -478,12 +478,67 @@ const trusted: JupyterFrontEndPlugin<void> = {
...
+                }
+              });
+            }
+          });
+        }
+      },
+    });
+
+    // Add `Edit Notebook Metadata` option to Edit menu
+    menu.editMenu.addItem({ type: 'command', command: id, rank: 40 });
```
Comment (jtpio) [inline packages/notebook-extension/src/index.ts:531]: How about adding the menu entry via the settings defined in the `schema` folder?

That way we can drop the dependency on `IMainMenu` in the plugin, and this lets user customize the menu entry more easily if they wish.

Comment (Dilip-Jain) [reply packages/notebook-extension/src/index.ts:531]: sure that does make sense @jtpio, but how do I do that.. 
with the examples I saw, I understood that I need to create a menus.json file in notebook-extension/schema folder and fill it with the menu emtry for command I created i.e. `notebook:edit-metadata`. But it didn't work. I also altered the index.ts file a little by exporting a const variable under namespace CommandIds.. but what after that!? I was really lost, so ended up using IMainMenu. 
any example you can point at? or maybe guide me with the steps and/or what I missed last time I tried??

Comment (jtpio) [reply packages/notebook-extension/src/index.ts:531]: An example would be the help extension, which defines the menu entries here:

https://github.com/jupyter/notebook/blob/e73d410074e6dbf97273f761d1513ff61db2965c/packages/help-extension/schema/open.json#L4-L19

With the commands being defined here:

https://github.com/jupyter/notebook/blob/e73d410074e6dbf97273f761d1513ff61db2965c/packages/help-extension/src/index.tsx#L40-L44

Comment (Dilip-Jain) [reply packages/notebook-extension/src/index.ts]: Hi @jtpio, the required changes have been made and pushed. please check, it should be ready to pull.

File: packages/notebook-extension/src/index.ts [hunk 3]
```diff
@@ -491,12 +498,66 @@ const trusted: JupyterFrontEndPlugin<void> = {
...
+
+        // Check if Show Notebook Tools (Right Sidebar) is open (expanded)
+        if (!commands.isToggled(command, args)) {
+          await commands.execute(command, args).then((_) => {
+            // For expanding the 'Advanced Tools' section (default: collapsed)
+            if (notebookTools) {
+              const tools = (notebookTools?.layout as any).widgets;
+              tools.forEach((tool: any) => {
+                if (
+                  tool.widget.title.label === trans.__('Advanced Tools') &&
```
Comment (jtpio) [inline packages/notebook-extension/src/index.ts:537]: This looks a little brittle. But I guess there is no other way to programmatically expand the "Advanced Tools" section?

Comment (Dilip-Jain) [reply packages/notebook-extension/src/index.ts:537]: that's true @jtpio, back while writing, I tried to find simpler and yk, more streamlined way of doing it but couldn't find any and functionally expanding 'Advanced Tools' section was necessary so couldn't avoid it.

Comment (jtpio) [reply packages/notebook-extension/src/index.ts:537]: Right, maybe we can stick to this for now then.

**Comment by jtpio** [comment]
FYI @Dilip-Jain I pushed a few commits to fix the menu item, and add the command to the command palette.

File: packages/notebook-extension/src/index.ts (hunk 3 above)
Comment (Dilip-Jain) [reply packages/notebook-extension/src/index.ts:542]: yes

**Comment by jtpio** [comment]
bot please update playwright snapshots

**Review by jtpio — APPROVED** [review]
Thanks!

**Pull request merged**